File: python/static-site/static_site/static_site_construct.py
from aws_cdk import (
    aws_cloudfront as cloudfront,
    aws_s3 as s3,
    aws_route53 as route53,
    aws_route53_targets as targets,
    aws_ssm as ssm,
    core,
)


class StaticSiteProps(object):

    def __init__(self, domain_name: str, site_sub_domain: str, hosted_zone_id: str):
        self._domain_name = domain_name
        self._site_sub_domain = site_sub_domain
        self._hosted_zone_id = hosted_zone_id

    @property
    def domain_name(self) -> str:
        return self._domain_name

    @property
    def site_sub_domain(self) -> str:
        return self._site_sub_domain

    @property
    def hosted_zone_id(self) -> str:
        return self._hosted_zone_id


class StaticSiteConstruct(core.Construct):
    @property
    def buckets(self):
        return tuple(self._buckets)

    def __init__(self, scope: core.Construct, id: str, props: StaticSiteProps) -> None:
        super().__init__(scope, id)

        site_domain: str = '{sub_domain}.{domain_name}'.format(
            sub_domain=props.site_sub_domain,
            domain_name=props.domain_name
        )

        # Content Bucket
        site_bucket = s3.Bucket(
            self,
            'SiteBucket',
            bucket_name=site_domain,
            website_index_document='index.html',
            website_error_document='error.html',
            public_read_access=True
        )
        core.CfnOutput(self, 'Bucket', value=site_bucket.bucket_name)

        # Pre-existing ACM Certificate, with ARN stored in an SSM Parameter
        certificate_arn: str = ssm.StringParameter.from_string_parameter_attributes(
            self,
            "MYCertArnString",
            parameter_name="CertificateArn-{}".format(site_domain)
        ).string_value

        # CloudFront distribution that provides HTTPS
        alias_configuration = cloudfront.AliasConfiguration(
            acm_cert_ref=certificate_arn,
            names=[site_domain],
            ssl_method=cloudfront.SSLMethod.SNI,
            security_policy=cloudfront.SecurityPolicyProtocol.TLS_V1_1_2016
        )

        source_configuration = cloudfront.SourceConfiguration(
            s3_origin_source=cloudfront.S3OriginConfig(
                s3_bucket_source=site_bucket
            ),
            behaviors=[cloudfront.Behavior(is_default_behavior=True)]
        )

        distribution = cloudfront.CloudFrontWebDistribution(
            self,
            'SiteDistribution',
            alias_configuration=alias_configuration,
            origin_configs=[source_configuration]
        )
        core.CfnOutput(self, 'DistributionId', value=distribution.distribution_id)

        # Route53 alias record for the CloudFront Distribution
        zone = route53.HostedZone.from_hosted_zone_attributes(
            self,
            id="HostedZoneID",
            hosted_zone_id=props.hosted_zone_id,
            zone_name=props.domain_name
        )

        route53.ARecord(
            self,
            'SiteAliasRecord',
            record_name=site_domain,
            target=route53.AddressRecordTarget.from_alias(targets.CloudFrontTarget(distribution)),
            zone=zone
        )

File: csharp/static-site/src/StaticSite/StaticSiteConstruct.cs
using Amazon.CDK;
using Amazon.CDK.AWS.Route53;
using Amazon.CDK.AWS.Route53.Targets;
using Amazon.CDK.AWS.S3;
using Amazon.CDK.AWS.S3.Deployment;
using Amazon.CDK.AWS.CertificateManager;
using Amazon.CDK.AWS.CloudFront;
using Constructs;
using System.Linq;

namespace StaticSite
{
    public class StaticSiteConstructProps
    {
        public string DomainName;

        public string SiteSubDomain;
    }

    public class StaticSiteConstruct : Construct
    {

        // A simple construct that contains a collection of AWS S3 buckets.
        public StaticSiteConstruct(Construct scope, string id, StaticSiteConstructProps props) : base(scope, id)
        {
            var zone = HostedZone.FromLookup(this, "Zone", new HostedZoneProviderProps
            {
                DomainName = props.DomainName
            });

            var siteDomain = (string) ($"{props.SiteSubDomain}.{props.DomainName}");
            new CfnOutput(this, "Site", new CfnOutputProps
            {
                Value = $"https://{siteDomain}"
            });

            var siteBucket = new Bucket(this, "SiteBucket", new BucketProps
            {
                BucketName = siteDomain,
                WebsiteIndexDocument = "index.html",
                WebsiteErrorDocument = "error.html",
                PublicReadAccess = true,

                // The default removal policy is RETAIN, which means that cdk destroy will not attempt to delete
                // the new bucket, and it will remain in your account until manually deleted. By setting the policy to
                // DESTROY, cdk destroy will attempt to delete the bucket, but will error if the bucket is not empty.
                RemovalPolicy = RemovalPolicy.DESTROY // NOT recommended for production code
            });

            new CfnOutput(this, "Bucket", new CfnOutputProps
            {
                Value = siteBucket.BucketName
            });

            var certificateArn = new DnsValidatedCertificate(this, "SiteCertificate", new DnsValidatedCertificateProps
            {
                DomainName = siteDomain,
                HostedZone = zone
            }).CertificateArn;

            new CfnOutput(this, "Certificate", new CfnOutputProps{Value = certificateArn});

            var behavior = new Behavior();
            behavior.IsDefaultBehavior = true;

            var distribution = new CloudFrontWebDistribution(this, "SiteDistribution", new CloudFrontWebDistributionProps
            {
                AliasConfiguration = new AliasConfiguration
                {
                    AcmCertRef = certificateArn,
                    Names = new string[] {siteDomain},
                    SslMethod = SSLMethod.SNI,
                    SecurityPolicy = SecurityPolicyProtocol.TLS_V1_2016
                },
                OriginConfigs = new ISourceConfiguration[]
                {
                    new SourceConfiguration
                    {
                        S3OriginSource = new S3OriginConfig
                        {
                            S3BucketSource = siteBucket
                        },
                        Behaviors = new Behavior[] {behavior}
                    }
                }
            });

            new CfnOutput(this, "DistributionId", new CfnOutputProps
            {
                Value = distribution.DistributionId
            });

            new ARecord(this, "SiteAliasRecord", new ARecordProps
            {
                RecordName = siteDomain,
                Target = RecordTarget.FromAlias(new CloudFrontTarget(distribution)),
                Zone = zone
            });

            new BucketDeployment(this, "DeployWithInvalidation", new BucketDeploymentProps
            {
                Sources = new ISource[] {Source.Asset("./site-contents")},
                DestinationBucket = siteBucket,
                Distribution = distribution,
                DistributionPaths = new string[] {"/*"}
            });
        }
    }
}
